server/sync.go

```diff
@@ -0,0 +1,223 @@
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/oklog/ulid"
+	"github.com/prometheus/prometheus/tsdb"
+	"golang.org/x/exp/slices"
+)
+
+type WriterFunc func(io.Reader, string) (int64, error)
+type ReaderFunc func(string) ([]byte, error)
+
+// syncFileStore does two periodic jobs and listens closeChan to exit from its loop.
+// 1. Synchronizes the local tsdb blocks with the remote filestore (mattermost-server)
+// 2. Periodically deletes the obsolete blocks from the remote filestore
+func (p *Plugin) syncFileStore() {
+	localStorageDir := p.db.Dir()
+	remoteStorageDir := filepath.Join(pluginDataDir, PluginName, tsdbDirName)
+
+	tickFileStoreSync := time.Tick(time.Duration(*p.configuration.FileStoreSyncPeriodMinutes) * time.Minute)
+	tickFileStoreCleanUp := time.Tick(time.Duration(*p.configuration.FileStoreCleanupPeriodMinutes) * time.Minute)
+
+loop:
+	for {
+		select {
+		case <-tickFileStoreSync:
+			p.API.LogDebug("Syncing blocks with the filestore")
+
+			entries, err := os.ReadDir(localStorageDir)
+			if err != nil {
+				p.API.LogError("could not list directories in local storage", "err", err)
+				break loop
+			}
+
+			// set the deadline for retention
+			ret := time.Now().AddDate(0, 0, -1**p.configuration.RetentionDurationDays)
+			blocksToSync := make([]string, 0)
+
+			for _, entry := range entries {
+				if !entry.IsDir() {
+					continue
+				}
+
+				if _, parseErr := ulid.Parse(entry.Name()); parseErr != nil {
+					// means that the directory is not a valid block
+					continue
+				}
+
+				f := filepath.Join(localStorageDir, entry.Name(), metaFileName)
+				meta, err2 := readBlockMeta(f, os.ReadFile)
+				if err2 != nil && (os.IsNotExist(err2) || errors.Is(err2, fs.ErrNotExist)) {
+					continue
+				} else if err2 != nil {
+					p.API.LogWarn("unable to read meta file", "err", err2)
+					continue
+				}
+
+				max := time.UnixMilli(meta.MaxTime)
+				if max.After(ret) {
+					blocksToSync = append(blocksToSync, entry.Name())
+				}
+			}
+
+			_, err = p.fileBackend.FileExists(PluginName)
+			if err != nil {
+				p.API.LogError("unable check file store, exiting sync loop", "err", err)
+				return
+			}
+
+			// we are using a pre-defined directory structure in the filestore
+			// hence using the same remoteStorageDir variable here
+			blocksInSFileStore := make(map[string]bool)
+			dirs, err2 := p.fileBackend.ListDirectory(remoteStorageDir)
+			if err2 != nil {
+				p.API.LogError("could not list directory from filestore", "err", err2)
+			}
+			for i := range dirs {
+				// we trim the parent dir from the dirs as filebackend returns the
+				// absolute path (relative to filestore root)
+				dir := strings.TrimPrefix(dirs[i], remoteStorageDir+string(os.PathSeparator))
```

Comment:
This https://pkg.go.dev/path/filepath#Base is what you want right?

Reply:
Genau!